hail/python/hail/ir/ir.py

```diff
@@ -1176,6 +1177,20 @@ def __eq__(self, other):
                other.init_op_args == self.init_op_args and \
                other.seq_op_args == self.seq_op_args
 
+    def __hash__(self):
+        h = hash(self.agg_op)
```

Comment:
This prompted me to take an interesting detour; wondering why 31 and 37; if I understood, these mimic the Java hash function (apparently 31 is common), the equivalent of << 5 -1 and I suppose <<5 + 5, and are used to more uniformly use the allocated space, to reduce risk of collision. 

https://stackoverflow.com/questions/299304/why-does-javas-hashcode-in-string-use-31-as-a-multiplier

Is that right, or is that explanation incomplete Tim?

Reply:
Usually when I write a hash function I just make a tuple of the things I care about and call python's hash on that. Would the results of doing that be worse than building your own hash function like this?

Reply:
that's fine, and probably easier! I may change it.

Reply:
So each individual hash already does something like this, but using 1000003 (http://effbot.org/zone/python-hash.htm)

If we're combining multiple hashes, do we get the same degree of entropy if we don't perform the shift in the addition, if our operands are hashes themselves?

Reply:
@akotlar - I don't know too much about the particulars of hashing (Patrick is the person for that!) but using primes in this way seems to be pretty standard.

XOR seems to be better than addition, too, which is what I was using.